opentelemetry-sdk/src/metrics/internal/mod.rs

```diff
@@ -164,35 +170,40 @@ where
     where
         MapFn: FnMut(Vec<KeyValue>, A) -> Res,
     {
-        prepare_data(dest, self.count.load(Ordering::SeqCst));
-        if self.has_no_attribute_value.swap(false, Ordering::AcqRel) {
+        // reset sorted trackers so new attributes set will be written into new hashmap
+        let trackers = match self.sorted_attribs.lock() {
+            Ok(mut trackers) => {                
+                let new = HashMap::with_capacity(trackers.len());
+                replace(trackers.deref_mut(), new)
```

Comment:
We have briefly touched this situation in my previous PR. [about unnecessary allocation](https://github.com/open-telemetry/opentelemetry-rust/pull/2267#discussion_r1828334335), but we didn't discussed downsides of it.
Here's the table of possible solutions.
| memory policy | approach                                                | benefits                                               | drawbacks                                                                                                                                                                                                                                                                                                                                                                                                                     |
|---------------|---------------------------------------------------------|--------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| concervative  | take(trackers)                                          | smallest memory footprint never grows more than needed | lots of allocation every time new collection phase starts                                                                                                                                                                                                                                                                                                                                                                     |
| balanced      | replace(trackers, HashMap with_capacity (previous_len)) | also small memory footprint                            | on stable loads might save a lot of unnecessary allocations, on spiky loads might either waist memory or do more allocations                                                                                                                                                                                                                                                                                                  |
| liberal       | replace(trackers, previous_tracker with same_memory)    | no allocations at all                                  | memory only grow, which might be a problem if you have huge spike, allocated memory will stay there forever (though it's only for contents of "empty" buckets, maybe not that bad?). Also it's a bit more complexity in code, as you need to have extra variable (and lock) just for that, and finally this might be unexpected for end user, as end user might expect that delta temporality doesn't do any sort of caching  |

At the moment I chose "balanced approach", I found it most natural :) but happy to hear your opinion on this too.

Reply:
@fraillt Please see https://github.com/open-telemetry/opentelemetry-rust/pull/2343 where I opted for the liberal from the above table. It should ensure HashMap starts with enough capacity, and never ever needs to re-adjust in its lifetime. Please take a look and share your feedback.

"_memory only grow_" - it will not, due to cardinality capping.
"_finally this might be unexpected for end user, as end user might expect that delta temporality doesn't do any sort of caching_" - these are internal changes only. end users should still see expected behavior with this change.